pkg/system/proxy/internal/test/proxy_test.go

```diff
@@ -0,0 +1,343 @@
+	}
+	defer conn.Close()
+
+	client := traits.NewMetadataApiClient(conn)
+	err = backoff.Retry(func() error {
+		_, err := client.GetMetadata(ctx, &traits.GetMetadataRequest{})
+		return err
+	}, backoff.WithContext(backoff.NewExponentialBackOff(), ctx))
+	if err != nil {
+		t.Fatalf("wait for node %s: %v", addr, err)
+	}
+}
+
+func configureCohort(t *testing.T, ctx context.Context) {
+	t.Helper()
+
+	// todo: use the hubs ca (should be in dir, after the first request) for our client cert checks
+
+	hubConn, err := grpc.DialContext(ctx, shared.HubGRPCAddr, grpc.WithTransportCredentials(credentials.NewTLS(&tls.Config{
+		InsecureSkipVerify: true,
+	})))
+	if err != nil {
+		t.Fatal("dial:", err)
+	}
+	defer hubConn.Close()
+
+	client := gen.NewHubApiClient(hubConn)
+	for _, addr := range shared.ACGRPCAddrs {
+		_, err := client.EnrollHubNode(ctx, &gen.EnrollHubNodeRequest{Node: &gen.HubNode{Address: addr}})
+		if err != nil {
+			t.Fatalf("enroll %s: %v", addr, err)
+		}
+	}
+	for _, addr := range shared.GWGRPCAddrs {
+		_, err := client.EnrollHubNode(ctx, &gen.EnrollHubNodeRequest{Node: &gen.HubNode{Address: addr}})
+		if err != nil {
+			t.Fatalf("enroll %s: %v", addr, err)
+		}
+	}
+}
+
+func testGW(t *testing.T, ctx context.Context, addr string) {
+	t.Helper()
+
+	conn, err := grpc.DialContext(ctx, addr, grpc.WithTransportCredentials(credentials.NewTLS(&tls.Config{
+		InsecureSkipVerify: true,
+	})))
+	if err != nil {
+		t.Fatalf("dial %s: %v", addr, err)
+	}
+	defer conn.Close()
+
+	deviceNames := []string{
+		"ac1/dev1",
+		"ac2/dev1",
+		"hub/dev1",
+	}
+	t.Logf("[%s] Waiting for cohort devices", addr)
+	for _, name := range deviceNames {
+		waitForDevice(t, ctx, conn, name)
+	}
+
+	t.Logf("[%s] Testing proxying", addr)
+	client := traits.NewOnOffApiClient(conn)
+	for _, name := range deviceNames {
+		testOnOffApi(t, ctx, addr, name, client)
+	}
+}
+
+func waitForDevice(t *testing.T, ctx context.Context, conn *grpc.ClientConn, name string) {
+	t.Helper()
+
+	client := traits.NewMetadataApiClient(conn)
+	err := backoff.Retry(func() error {
+		_, err := client.GetMetadata(ctx, &traits.GetMetadataRequest{Name: name})
+		return err
+	}, backoff.WithContext(backoff.NewExponentialBackOff(), ctx))
+	if err != nil {
+		t.Fatalf("wait for device %s: %v", name, err)
+	}
+}
+
+func testOnOffApi(t *testing.T, ctx context.Context, addr, name string, client traits.OnOffApiClient) {
+	t.Helper()
+
+	// useful for cancelling the stream
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	// set initial known state: ON
+	{
+		res, err := client.UpdateOnOff(ctx, &traits.UpdateOnOffRequest{Name: name, OnOff: &traits.OnOff{State: traits.OnOff_ON}})
+		if err != nil {
+			t.Fatalf("[%s] update onoff %s: %v", addr, name, err)
+		}
+		if diff := cmp.Diff(res, &traits.OnOff{State: traits.OnOff_ON}, protocmp.Transform()); diff != "" {
+			t.Fatalf("[%s] update onoff %s: unexpected response (-want +got):\n%s", addr, name, diff)
+		}
+	}
```

Comment:
I understand what you mean. I mostly did it to avoid the different `res` values needing to be the same type (though I think they are at that scope level).

Reply:
I'll get rid of the blocks, they weren't adding anything anymore anyway